File: python/include/print_util.py
'''
File: print_util.py
Created Time: 2021-12-11
Author: Krahets (krahets@163.com)
'''

from .binary_tree import TreeNode, tree_to_list
from .linked_list import ListNode, linked_list_to_list

# Print a matrix
def print_matrix(mat):
    pstr = []
    for arr in mat:
        pstr.append('  ' + str(arr))

    print('[\n' + ',\n'.join(pstr) + '\n]')

# Print a linked list
def print_linked_list(head):
    arr = linked_list_to_list(head)
    print(' -> '.join([str(a) for a in arr]))

# Print a binary tree (90º counter-clockwise rotated)
def print_tree(root):
    def helper(root, level):
        if not root:
            return
        helper(root.right, level + 1)
        print(' ' * 4 * level + '->', root.val)
        helper(root.left, level + 1)
    helper(root, 0)

File: java/include/PrintUtil.java
package include;

import java.util.*;

public class PrintUtil {
    /**
     * Print a linked list
     * @param head
     */
    public static void printLinkedList(ListNode head) {
        List<String> list = new ArrayList<>();
        while (head != null) {
            list.add(String.valueOf(head.val));
            head = head.next;
        }
        System.out.println(String.join(" -> ", list));
    }

    /**
     * Print a binary tree (90º counter-clockwise rotated)
     * @param root
     */
    public static void printTree(TreeNode root) {
        printTreeHelper(root, 0);
    }

    /**
     * Print helper for binary tree
     * @param root
     * @param level
     */
    private static void printTreeHelper(TreeNode root, int level) {
        if (root == null)
            return;
        printTreeHelper(root.right, level + 1);
        System.out.println(" ".repeat(4 * level) + "->" + root.val);
        printTreeHelper(root.left, level + 1);
    }
}
